src/lib/operators/aggregate_sort.cpp

```diff
@@ -0,0 +1,585 @@
+ * @param aggregate_results the vector where the aggregate values should be stored
+ * @param aggregate_null_values the vector indicating whether a specific aggregate value is null
+ * @param aggregate_group_index the offset to use for <code>aggregate_results</code> and <code>aggregate_null_values</code>
+ * @param aggregate_index current aggregate's offset in <code>_aggregates</code>
+ * @param current_aggregate_value the value of the aggregate (return value of the aggregate function) - used by all except COUNT (all versions)
+ * @param value_count the number of non-null values - used by COUNT(<name>) and AVG
+ * @param value_count_with_null the number of rows  - used by COUNT(*)
+ * @param unique_value_count the number of unique values
+ */
+template <typename AggregateType, AggregateFunction function>
+void AggregateSort::_set_and_write_aggregate_value(
+    std::vector<AggregateType>& aggregate_results, std::vector<bool>& aggregate_null_values,
+    const uint64_t aggregate_group_index, [[maybe_unused]] const uint64_t aggregate_index,
+    std::optional<AggregateType>& current_aggregate_value, [[maybe_unused]] const uint64_t value_count,
+    [[maybe_unused]] const uint64_t value_count_with_null, [[maybe_unused]] const uint64_t unique_value_count) const {
+  if constexpr (function == AggregateFunction::Count) {  // NOLINT
+    if (this->_aggregates[aggregate_index].column) {
+      // COUNT(<name>), so exclude null values
+      current_aggregate_value = value_count;
+    } else {
+      // COUNT(*), so include null values
+      current_aggregate_value = value_count_with_null;
+    }
+  }
+  if constexpr (function == AggregateFunction::Avg && std::is_arithmetic_v<AggregateType>) {  // NOLINT
+    // this ignores the case of Avg on strings, but we check in _on_execute() this does not happen
+
+    if (value_count == 0) {
+      // there are no non-null values, the average itself must be null (otherwise division by 0)
+      current_aggregate_value = std::optional<AggregateType>();
+    } else {
+      // normal average calculation
+      current_aggregate_value = *current_aggregate_value / value_count;
+    }
+  }
+  if constexpr (function == AggregateFunction::CountDistinct) {  // NOLINT
+    current_aggregate_value = unique_value_count;
+  }
+
+  // check if the value is a null value
+  aggregate_null_values[aggregate_group_index] = !(current_aggregate_value.has_value());
+  if (current_aggregate_value.has_value()) {
+    // only store non-null values
+    aggregate_results[aggregate_group_index] = *current_aggregate_value;
+  }
+}
+
+/**
+ * Executes the aggregation.
+ * High-level overview:
+ *
+ * Make sure the aggregates are valid
+ * Build the output table definition
+ * if empty
+ *   return (empty) result table
+ *
+ * Sort the input table after all group by columns.
+ *  Currently, this is done using multiple passes of the Sort operator (which is stable)
```

Comment:
Add a comment that future implementations might pass in sorted data?

Reply:
Related to #1519, should the optimizer handle that?

Reply:
I have referenced the issue and added a short comment.
I would suggest to also add a (more detailed) comment on interesting orders/sortedness/consecutiveness in the top-level comment for the aggregate (probably aggregate_sort.hpp)